# Tarea 1: Introducción, Vector Space Models y Information Retrieval </h1>
**Procesamiento de Lenguaje Natural (CC6205-1 - Otoño 2024)**

## Tarjeta de identificación

**Nombres:** ```<Completar aquí>```

**Fecha límite de entrega 📆:** 09/04.

**Tiempo estimado de dedicación:** 4 horas


## Instrucciones
Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en **Information Retrieval (IR)** y **Vector Space Models**. Si aún no has visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP.

* La tarea es en **grupo** (maximo hasta 3 personas).
* La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
* El formato de entrega es este mismo Jupyter Notebook.
* Al momento de la revisión su código será ejecutado. Por favor verifiquen que su entrega no tenga errores de compilación.
* Completar la tarjeta de identificación. Sin ella no podrá tener nota.

## Material de referencia

Diapositivas del curso 📄
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos del curso 📺

- [CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

## P1. Tokenización

En el primer ejercicio veremos la dificultad de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo.

In [1]:
# En caso de desarrollar la tarea desde colab, con el siguiente código podemos cargar los archivos desde drive:

try:
    from google.colab import drive

    drive.mount("/content/drive", force_remount=True)
    path = '/content/drive/MyDrive/nlp/oh_algoritmo.txt'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook


In [3]:
try:
    # Abre el archivo en modo lectura ("r")
    with open("oh_algoritmo.txt", "r") as archivo:
        # Lee el contenido del archivo
        texto = archivo.read()
        # Imprime el contenido
        print(texto)
except FileNotFoundError:
    print("El archivo no se encuentra.")
except Exception as e:
    print("Ocurrió un error:", e)

[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]

[Refrán: Jorge Drexler]
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?

[Estribillo: Jorge Drexler]
Dime qué debo cantar
Oh, algoritmo
Sé que lo sabes mejor
Incluso que yo mismo

[Verso 1: Nora Erez]
Wait, what's that money that you spent?
What's that sitting on your plate?
Do you want what you've been fed?
Are you the fish or bait?
Mmm, I'm on the top of the roof and I feel like a jail
Rather not pay the bail
To dangerous people with blood on their faces
So I'm sharing a cell with the masses
The underground always strive for the main
Streaming like Grande's big-ass ring
Screaming: I'll write you out my will
Conscious is free, but not the will
Conscious is free, but not the will
You

Fuente: https://genius.com/Jorge-drexler-oh-algoritmo-lyrics

### Pregunta 1.a (0.25 puntos)

Diseñe una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Es libre de elegir la forma de tokenizar mientras no utilice librerías con tokenizadores ya implementados. Puede utilizar la librería **re** importada para trabajar símbolos. Explique su razonamiento.


In [4]:
import re

In [5]:
def get_tokens(texto):
    # Definimos el patrón de expresión regular para encontrar tokens
    patron = r'\b\w+\b'  # Este patrón busca palabras completas

    # Utilizamos la función findall de re para encontrar todas las coincidencias del patrón en el texto
    tokens = re.findall(patron, texto)

    return tokens

In [6]:
tokens = get_tokens(texto)
tokens

['Letra',
 'de',
 'Oh',
 'Algoritmo',
 'ft',
 'Nora',
 'Erez',
 'Refrán',
 'Jorge',
 'Drexler',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 'Estribillo',
 'Jorge',
 'Drexler',
 'Dime',
 'qué',
 'debo',
 'cantar',
 'Oh',
 'algoritmo',
 'Sé',
 'que',
 'lo',
 'sabes',
 'mejor',
 'Incluso',
 'que',
 'yo',
 'mismo',
 'Verso',
 '1',
 'Nora',
 'Erez',
 'Wait',
 'what',
 's',
 'that',
 'money',
 'that',
 'you',
 'spent',
 'What',
 's',
 'that',
 'sitting',
 'on',
 'your',
 'plate',
 'Do',
 'you',
 'want',


### Pregunta 1.b (0.25 puntos)
Explique su implementación aquí:
> La implementación ...

Implementación con la libreria NLTK

In [7]:
from nltk.tokenize import wordpunct_tokenize
nltk_tokens = wordpunct_tokenize(texto)
nltk_tokens

['[',
 'Letra',
 'de',
 '"¡',
 'Oh',
 ',',
 'Algoritmo',
 '!"',
 'ft',
 '.',
 'Nora',
 'Erez',
 ']',
 '[',
 'Refrán',
 ':',
 'Jorge',
 'Drexler',
 ']',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '¿',
 'Quién',
 'quiere',
 'que',
 'yo',
 'quiera',
 'lo',
 'que',
 'creo',
 'que',
 'quiero',
 '?',
 '[',
 'Estribillo',
 ':',
 'Jorge',
 'Drexler',
 ']',
 'Dime',
 'qué',
 'debo',
 'cantar',
 'Oh',
 ',',
 'algoritmo',
 'Sé',
 'que',
 'lo',
 'sabes',
 'mejor',
 'Incluso',
 'que',
 'yo',
 'mismo',
 '[',
 'Verso',
 '1',
 ':',
 'Nora',
 

### Pregunta 1.c (0.5 puntos)
¿Qué diferencias y similitudes encontrase al comparar la función de tokenización creada manualmente por ti contra la implementación de NLTK, al tokenizar la letra de la canción "Oh, algoritmo"?

> Escribe tu respuesta aquí

## P2. Stemming y Stopwords

En esta sección debera implementar funciones de stemming y stopwords basado en lo visto en clase. En la siguiente celda tiene el corpus que usara en esta sección:

In [8]:
# Corpus en español
corpus_espanol = [
    "¿Quién quiere que yo quiera lo que creo que quiero?",
    "Dime qué debo cantar",
    "Sé que lo sabes mejor"
]

# Corpus en inglés
corpus_ingles = [
    "What's that sitting on your plate?",
    "Do you want what you've been fed?",
    "Are you the fish or bait?"
]

### Pregunta 2.a (0.5 puntos)
Implemente una función **`get_vocab()`** que extraiga los tokens de un corpus. Puede utilizar la función de la sección anterior.

In [9]:
def get_vocab(corpus):
  ### Aquí inicia tu código ###

  tokens = []
  for sent in corpus:
    tokens += get_tokens(sent)
  return list(set(tokens))

  ### Aquí termina tu código ###

In [10]:
vocab_espanol = get_vocab(corpus_espanol)
vocab_espanol

['debo',
 'Dime',
 'Sé',
 'mejor',
 'qué',
 'yo',
 'lo',
 'sabes',
 'quiere',
 'creo',
 'que',
 'Quién',
 'quiero',
 'quiera',
 'cantar']

Resultado esperado (el orden puede variar):
```
['yo', 'debo', 'creo', 'Dime', 'lo', 'cantar', 'mejor', 'Sé', 'que', 'quiere', 'quiero', 'sabes', 'Quién', 'quiera', 'qué']
```

In [11]:
vocab_ingles = get_vocab(corpus_ingles)
vocab_ingles

['fish',
 'you',
 'on',
 'bait',
 've',
 'been',
 'your',
 'that',
 's',
 'or',
 'plate',
 'the',
 'What',
 'Are',
 'sitting',
 'Do',
 'want',
 'what',
 'fed']

Resultado esperado:
```
['fed', 'been', 'or', 'want', 'plate', 'the', 've', 'your', 's', 'you', 'what', 'Are', 'bait', 'What', 'fish', 'that', 'sitting', 'Do', 'on']
```

### Pregunta 2.b (0.5 puntos)
Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario preprocesado. Explique las reglas de stemming y elección de stopwords:

    Explique sus reglas aquí:


In [13]:
def pre_processing(vocabulario, idioma):
    ### Aquí inicia tu código ###
    def stem_espanol(word):
      # Reglas de sustitución para el algoritmo de Porter en español
      substitutions = [
          ('a', ''),   # Elimina la terminación 'a'
          ('o', ''),   # Elimina la terminación 'o'
          ('á', 'a'),  # Reemplaza 'á' con 'a'
          ('é', 'e'),  # Reemplaza 'é' con 'e'
          ('í', 'i'),  # Reemplaza 'í' con 'i'
          ('ó', 'o'),  # Reemplaza 'ó' con 'o'
          ('ú', 'u'),  # Reemplaza 'ú' con 'u'
          ('e', ''),   # Elimina la terminación 'e'
          ('é', ''),   # Elimina la terminación 'é'
          ('ar', ''),  # Elimina la terminación 'ar'
          ('es', 'e')  # Reemplaza 'es' con 'e'
      ]

      for old, new in substitutions:
          if word.endswith(old):
              return word[:-len(old)] + new

      return word

    def stem_ingles(word):
      # Reglas de sustitución para el algoritmo de Porter en inglés
      substitutions = [
          ('sses', 'ss'),  # Reemplaza 'sses' con 'ss'
          ('ies', 'i'),    # Reemplaza 'ies' con 'i'
          ('s', ''),       # Elimina la terminación 's'
          ('ed', ''),      # Elimina la terminación 'ed'
          ('ing', ''),     # Elimina la terminación 'ing'
          ('y', 'i'),      # Reemplaza 'y' con 'i'
      ]

      for old, new in substitutions:
          if word.endswith(old):
              return word[:-len(old)] + new

      return word

    stopwords_espanol = ["que", "yo", "lo", "qué", "debo", "sé", "que", "lo"]
    stopwords_ingles = ["do", "fed", "been", "or", "the", "ve", "your", "s", "you", "what", "are", "that", "on"]

    if idioma == 'espanol':
        stopwords = stopwords_espanol
        stemmer = stem_espanol
    elif idioma == 'ingles':
        stopwords = stopwords_ingles
        stemmer = stem_ingles
    else:
        raise ValueError("Idioma no válido")

    vocab_procesado = [stemmer(palabra) for palabra in vocabulario if palabra.lower() not in stopwords]

    return vocab_procesado
    ### Aquí termina tu código ###

In [14]:
# Aplicar preprocesamiento a los vocabularios de ejemplo con NLTK
vocab_procesado_espanol = pre_processing(vocab_espanol, 'espanol')
vocab_procesado_ingles = pre_processing(vocab_ingles, 'ingles')

# Mostrar resultados
print("Vocabulario procesado en español:", vocab_procesado_espanol, "\n")
print("Vocabulario procesado en inglés:", vocab_procesado_ingles, "\n")

Vocabulario procesado en español: ['Dim', 'mejor', 'sabe', 'quier', 'cre', 'Quién', 'quier', 'quier', 'cant'] 

Vocabulario procesado en inglés: ['fish', 'bait', 'plate', 'sitt', 'want'] 



## P3. Bag of Words (0.5 puntos)
Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:


In [15]:
d0 = 'El pájaro come semillas'
d1 = 'El pájaro se despierta y canta'
d2 = 'El pájaro canta y come semillas'
d3 = 'El pez come y nada en el agua'
d4 = 'El pez empieza a nadar'
d5 = 'El pez come alimento'
corpus = [d0, d1, d2, d3, d4, d5]

El objetivo da las siguientes secciones es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica **TF-IDF**.

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es la de **Bag of Words**, método mediante el cual se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que recibe como input un arreglo de documentos y devuelve un dataframe de pandas con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representan las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el BoW de un documento.

***Disclaimer: el orden de los resultados pueden variar.***


Por ejemplo para el siguiente corpus:

```
corpus = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |




In [16]:
import pandas as pd

Implementar función `bag_of_words()`

In [17]:
def bag_of_words(corpus):
  ### Aquí inicia tu código ###
  all_bow = []
  for document in corpus:
      document_bow = {word: 0 for word in get_vocab(corpus)}
      for word in get_tokens(document):
          document_bow[word] += 1
      all_bow.append(document_bow)
  return pd.DataFrame(all_bow, index = [f"d{i}" for i in range(len(corpus))])
  ### Aquí termina tu código ###

In [18]:
dataset_bow = bag_of_words(corpus)
dataset_bow

despierta  come  pájaro  El  y  a  nada  en  empieza  alimento  pez  \
d0          0     1       1   1  0  0     0   0        0         0    0   
d1          1     0       1   1  1  0     0   0        0         0    0   
d2          0     1       1   1  1  0     0   0        0         0    0   
d3          0     1       0   1  1  0     1   1        0         0    1   
d4          0     0       0   1  0  1     0   0        1         0    1   
d5          0     1       0   1  0  0     0   0        0         1    1   

    canta  se  el  agua  semillas  nadar  
d0      0   0   0     0         1      0  
d1      1   1   0     0         0      0  
d2      1   0   0     0         1      0  
d3      0   0   1     1         0      0  
d4      0   0   0     0         0      1  
d5      0   0   0     0         0      0

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

## P4. TF-IDF

### 4.a TF (0.25 puntos)

Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia $\max_i({\text{tf}_{i,j}})$, donde
$i$ corresponde al índice de las filas (BoW) y $j$ al de las columnas (palabras). Es decir, dividir cada BoW sobre la cantidad de veces de la palabra que aparezca más veces en ese vector.


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{\max_i({\text{tf}_{i,j})}}$$

Implemente la función `calc_tf(dataset_bow)`, que entrega la matriz de TF normalizada del BoW del dataset.

In [19]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    return dataset_bow.divide(dataset_bow.max(axis=1), axis=0)
    ### Aquí termina tu código ###

In [20]:
tf = calc_tf(dataset_bow)
tf

despierta  come  pájaro   El    y    a  nada   en  empieza  alimento  pez  \
d0        0.0   1.0     1.0  1.0  0.0  0.0   0.0  0.0      0.0       0.0  0.0   
d1        1.0   0.0     1.0  1.0  1.0  0.0   0.0  0.0      0.0       0.0  0.0   
d2        0.0   1.0     1.0  1.0  1.0  0.0   0.0  0.0      0.0       0.0  0.0   
d3        0.0   1.0     0.0  1.0  1.0  0.0   1.0  1.0      0.0       0.0  1.0   
d4        0.0   0.0     0.0  1.0  0.0  1.0   0.0  0.0      1.0       0.0  1.0   
d5        0.0   1.0     0.0  1.0  0.0  0.0   0.0  0.0      0.0       1.0  1.0   

    canta   se   el  agua  semillas  nadar  
d0    0.0  0.0  0.0   0.0       1.0    0.0  
d1    1.0  1.0  0.0   0.0       0.0    0.0  
d2    1.0  0.0  0.0   0.0       1.0    0.0  
d3    0.0  0.0  1.0   1.0       0.0    0.0  
d4    0.0  0.0  0.0   0.0       0.0    1.0  
d5    0.0  0.0  0.0   0.0       0.0    0.0

Solución esperada:

|    |   El |   pájaro |   despierta |   el |   come |   a |   nadar |   se |   en |   y |   alimento |   semillas |   pez |   empieza |   canta |   agua |   nada |
|:---|-----:|---------:|------------:|-----:|-------:|----:|--------:|-----:|-----:|----:|-----------:|-----------:|------:|----------:|--------:|-------:|-------:|
| d0 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          0 |          1 |     0 |         0 |       0 |      0 |      0 |
| d1 |    1 |        1 |           1 |    0 |      0 |   0 |       0 |    1 |    0 |   1 |          0 |          0 |     0 |         0 |       1 |      0 |      0 |
| d2 |    1 |        1 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   1 |          0 |          1 |     0 |         0 |       1 |      0 |      0 |
| d3 |    1 |        0 |           0 |    1 |      1 |   0 |       0 |    0 |    1 |   1 |          0 |          0 |     1 |         0 |       0 |      1 |      1 |
| d4 |    1 |        0 |           0 |    0 |      0 |   1 |       1 |    0 |    0 |   0 |          0 |          0 |     1 |         1 |       0 |      0 |      0 |
| d5 |    1 |        0 |           0 |    0 |      1 |   0 |       0 |    0 |    0 |   0 |          1 |          0 |     1 |         0 |       0 |      0 |      0 |

### 4.b IDF (0.5 puntos)

Implementar `calc_idf(dataset_bow)`. Ésta debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el cálculo de cada idf por palabra.

Recordar que $\text{idf}_{t_i} = \log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ número de documentos que contienen la palabra $t_i$

In [21]:
import numpy as np

In [22]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    number_of_docs = len(dataset_bow)
    idf = {}
    for word in dataset_bow:
        n_i = (dataset_bow[[word]] != 0).sum()[0]
        idf[word] = np.log10(number_of_docs/n_i)
    return idf
    ### Aquí termina tu código ###

In [23]:
idf = calc_idf(dataset_bow)
idf

{'despierta': 0.7781512503836436,
 'come': 0.17609125905568124,
 'pájaro': 0.3010299956639812,
 'El': 0.0,
 'y': 0.3010299956639812,
 'a': 0.7781512503836436,
 'nada': 0.7781512503836436,
 'en': 0.7781512503836436,
 'empieza': 0.7781512503836436,
 'alimento': 0.7781512503836436,
 'pez': 0.3010299956639812,
 'canta': 0.47712125471966244,
 'se': 0.7781512503836436,
 'el': 0.7781512503836436,
 'agua': 0.7781512503836436,
 'semillas': 0.47712125471966244,
 'nadar': 0.7781512503836436}

Solución esperada:
```
{'El': 0.0,
 'pájaro': 0.3010299956639812,
 'despierta': 0.7781512503836436,
 'el': 0.7781512503836436,
 'come': 0.17609125905568124,
 'a': 0.7781512503836436,
 'nadar': 0.7781512503836436,
 'se': 0.7781512503836436,
 'en': 0.7781512503836436,
 'y': 0.3010299956639812,
 'alimento': 0.7781512503836436,
 'semillas': 0.47712125471966244,
 'pez': 0.3010299956639812,
 'empieza': 0.7781512503836436,
 'canta': 0.47712125471966244,
 'agua': 0.7781512503836436,
 'nada': 0.7781512503836436}
 ```

### 4.c TF-IDF (0.25 puntos)
Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [24]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###

    return tf * idf

    ### Aquí termina tu código ###

In [25]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

despierta      come   pájaro   El        y         a      nada        en  \
d0   0.000000  0.176091  0.30103  0.0  0.00000  0.000000  0.000000  0.000000   
d1   0.778151  0.000000  0.30103  0.0  0.30103  0.000000  0.000000  0.000000   
d2   0.000000  0.176091  0.30103  0.0  0.30103  0.000000  0.000000  0.000000   
d3   0.000000  0.176091  0.00000  0.0  0.30103  0.000000  0.778151  0.778151   
d4   0.000000  0.000000  0.00000  0.0  0.00000  0.778151  0.000000  0.000000   
d5   0.000000  0.176091  0.00000  0.0  0.00000  0.000000  0.000000  0.000000   

     empieza  alimento      pez     canta        se        el      agua  \
d0  0.000000  0.000000  0.00000  0.000000  0.000000  0.000000  0.000000   
d1  0.000000  0.000000  0.00000  0.477121  0.778151  0.000000  0.000000   
d2  0.000000  0.000000  0.00000  0.477121  0.000000  0.000000  0.000000   
d3  0.000000  0.000000  0.30103  0.000000  0.000000  0.778151  0.778151   
d4  0.778151  0.000000  0.30103  0.000000  0.000000  0.000000  0.000000   
d5  0.000000  0.778151  0.30103  0.000000  0.000000  0.000000  0.000000   

    semillas     nadar  
d0  0.477121  0.000000  
d1  0.000000  0.000000  
d2  0.477121  0.000000  
d3  0.000000  0.000000  
d4  0.000000  0.778151  
d5  0.000000  0.000000

Solución esperada:

|    |   El |   pájaro |   despierta |       el |     come |        a |    nadar |       se |       en |       y |   alimento |   semillas |     pez |   empieza |    canta |     agua |     nada |
|:---|-----:|---------:|------------:|---------:|---------:|---------:|---------:|---------:|---------:|--------:|-----------:|-----------:|--------:|----------:|---------:|---------:|---------:|
| d0 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0        |   0.477121 | 0       |  0        | 0        | 0        | 0        |
| d1 |    0 |  0.30103 |    0.778151 | 0        | 0        | 0        | 0        | 0.778151 | 0        | 0.30103 |   0        |   0        | 0       |  0        | 0.477121 | 0        | 0        |
| d2 |    0 |  0.30103 |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0.30103 |   0        |   0.477121 | 0       |  0        | 0.477121 | 0        | 0        |
| d3 |    0 |  0       |    0        | 0.778151 | 0.176091 | 0        | 0        | 0        | 0.778151 | 0.30103 |   0        |   0        | 0.30103 |  0        | 0        | 0.778151 | 0.778151 |
| d4 |    0 |  0       |    0        | 0        | 0        | 0.778151 | 0.778151 | 0        | 0        | 0       |   0        |   0        | 0.30103 |  0.778151 | 0        | 0        | 0        |
| d5 |    0 |  0       |    0        | 0        | 0.176091 | 0        | 0        | 0        | 0        | 0       |   0.778151 |   0        | 0.30103 |  0        | 0        | 0        | 0        |

## P5. Cosine-similarity (0.25 puntos)
Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica.

Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos. Concluya cuáles son los dos documentos más similares.

In [26]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###

  return np.dot(v1,v2)/(np.linalg.norm(v1)*np.linalg.norm(v2))

  ### Aquí termina tu código ###

In [ ]:
similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity

for i in range(6):
  mask = [k != i for k in range(6)]
  j = np.argmax(similarity_matrix[i][mask])

  print(corpus[i])
  print("> Mas similar:", np.array(corpus)[mask][j])
  print("> Similitud:", similarity_matrix[i][mask][j], "\n")

El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520414 

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.39320101823128945 

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520414 

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406168 

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.07695078406752713 

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.0878790037217323 



Solución esperada:
```
El pájaro come semillas
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.7233435041520414

El pájaro se despierta y canta
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.39320101823128945

El pájaro canta y come semillas
> Mas similar: El pájaro come semillas
> Similitud: 0.7233435041520414

El pez come y nada en el agua
> Mas similar: El pájaro canta y come semillas
> Similitud: 0.09171890791406168

El pez empieza a nadar
> Mas similar: El pez come alimento
> Similitud: 0.07695078406752713

El pez come alimento
> Mas similar: El pez come y nada en el agua
> Similitud: 0.0878790037217323
```

## P6 N-gramas (1 punto)

En esta sección debera determinar los n-gramas del la cancion "Oh algoritmo".

### 6.a Corpus de entrenamiento y test (0.25 puntos)

En esta subsección debera definir el conjunto de entrenamiento y test de un corpus. Eliga una particion del 80% y 20% del texto.

In [27]:
try:
    # Abre el archivo en modo lectura ("r")
    with open("oh_algoritmo.txt", "r") as archivo:
        # Lee el contenido del archivo
        texto = archivo.read()
        # Imprime el contenido
        print(texto)
except FileNotFoundError:
    print("El archivo no se encuentra.")
except Exception as e:
    print("Ocurrió un error:", e)

[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]

[Refrán: Jorge Drexler]
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?
¿Quién quiere que yo quiera lo que creo que quiero?

[Estribillo: Jorge Drexler]
Dime qué debo cantar
Oh, algoritmo
Sé que lo sabes mejor
Incluso que yo mismo

[Verso 1: Nora Erez]
Wait, what's that money that you spent?
What's that sitting on your plate?
Do you want what you've been fed?
Are you the fish or bait?
Mmm, I'm on the top of the roof and I feel like a jail
Rather not pay the bail
To dangerous people with blood on their faces
So I'm sharing a cell with the masses
The underground always strive for the main
Streaming like Grande's big-ass ring
Screaming: I'll write you out my will
Conscious is free, but not the will
Conscious is free, but not the will
You

Defina una funcion `get_sentences()` que entregue todas las frases del corpus. Todas las frase con al menos una palabras.

In [28]:
def get_sentences(texto):
  frases = texto.split("\n")

  frases_limpias = []
  for f in frases:
    if len(f.strip()) > 0:
      frases_limpias.append(f)

  return frases_limpias

In [29]:
frases_limpias = get_sentences(texto)
frases_limpias

['[Letra de "¡Oh, Algoritmo!" ft. Nora Erez]',
 '[Refrán: Jorge Drexler]',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '¿Quién quiere que yo quiera lo que creo que quiero?',
 '[Estribillo: Jorge Drexler]',
 'Dime qué debo cantar',
 'Oh, algoritmo',
 'Sé que lo sabes mejor',
 'Incluso que yo mismo',
 '[Verso 1: Nora Erez]',
 "Wait, what's that money that you spent?",
 "What's that sitting on your plate?",
 "Do you want what you've been fed?",
 'Are you the fish or bait?',
 "Mmm, I'm on the top of the roof and I feel like a jail",
 'Rather not pay the bail',
 'To dangerous people with blood on their faces',
 "So I'm sharing a cell with the masses",
 'The underground always strive for the main',
 "Streaming like Grande's big-ass ring",
 "Screaming: I'll wr

Deria obtener en totasl 87 frases.

In [30]:
len(frases_limpias) == 87

True

Ahora definiremos el conjunto de entrenamiento y preuba para las frases:

In [31]:
split = int(len(frases_limpias) * 0.8)
train_corpus = frases_limpias[:split]
test_corpus = frases_limpias[split:]

len(train_corpus), len(test_corpus)

(69, 18)

### 6.b Estimación de N-gramas (0.5 puntos)

Defina una función que reciba una lista de frases de un corpus y un N que indique el tamaño de los N-gramas. La función debe retornar un diccionario de Python donde la llave es un token (o palabra) y el valor es la cantidad de veces que ocurre el token, es decir, la frecuencia. En el caso de N-gramas con N mayor a 1 (como bi-gramas o tri-gramas) debe añadir un token especial al inicio o final de cada frase según corresponda (ver clases del curso).

In [32]:
from nltk.tokenize import word_tokenize

In [33]:
def n_grams(corpus, n=3):
  ## Implementar aqui
  pass
  ##


In [34]:
n_grams(train_corpus, n=1)

In [35]:
n_grams(train_corpus, n=2)

In [36]:
n_grams(train_corpus, n=3)

Debe mostrar que su método funciona para $N$ = 1,2,3.

## P7. Perplexity (1 punto)

En esta sección evaluarán su modelo de n-gramas y determinarán la probabilidad de frases y la perplejidad con un conjunto de test.

### 7.a Obtener probabilidades (0.5 puntos)

En esta sección implementara una función que determine la probabilidad de una frase.

Defina una función que reciba una frase, un diccionario con n-gramas y el valor de $n$. La función debe entregar la probabilidad de cualquier frase.

In [ ]:
def get_probability(sentence, n_grams_frequency, n):
  ## Implementar aquí
  pass
  ##

Pruebe su función con frases frecuentes y comente sus resultados



### 7.b Perplexity en conjunto de test (0.5 puntos)

En esta sub-sección deberá calcular la perplejidad del corpus de test.

Defina una función que reciba un corpus de test y retorne la perplexity (ver clases del curso). Utilize la función de la sección anterior.

In [ ]:
def get_perplexity(corpus, n):
  ## implementar aqui
  pass
  ##

In [ ]:
get_perplexity(test_corpus, 3)

Dé una interpretacion de la perplexity en el corpus de test:
```
Nosotros ...
```

## P8. Interpolación Lineal (0.5 puntos)

Cree una función que obtenga la probabilidad de una frase interpolando linealmente modelos de unigrama, bigrama y trigrama ponderados por $\lambda_1, \lambda_2$ y $\lambda_3$ respectivamente. Para esto use las funciones que creó anteriormente.

In [ ]:
def get_probability_lineal_interpol(sentence, corpus, l_1, l_2, l_3):
  pass

Defina una función para calcular la perplejidad de un corpus con interpolación lineal.

In [ ]:
def get_pp_interpol(corpus, l_1, l_2, l_3):
  pass

Ahora haga pruebas con distintos valores de $\lambda_1, \lambda_2$ y $\lambda_3$, incluyendo valores extremos (por ejemplo $[1, 0, 0]$). Comente sus resultados.